framework: allow skipping tests based on wgslLanguageFeatures

dneto0 opened the pull request:
Issue: #3057

This is in preparation for #3056 

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

dneto0 reviewed (commented):
I haven't tested the case where this actually skips the test.

dneto0 commented on src/webgpu/util/device_pool.ts:
```diff
@@ -296,9 +307,13 @@ class DeviceHolder implements DeviceProvider {
+    langFeature: WGSLLanguageFeature | undefined
   ): Promise<DeviceHolder> {
     const gpu = getGPU(recorder);
+    if (langFeature && gpu.wgslLanguageFeatures.has(langFeature)) {
```
I haven't checked that this triggers.

kainino0x reviewed (commented):
It's surprising to me to wire wgslLanguageFeatures through DeviceSelectionDescriptor when it's never needed for device selection. I'm not sure how this is intended to be used - could we just have a `requireLanguageFeatureOrSkipTestCase` (that just calls `this.skip()` alongside `selectDeviceOrSkipTestCase` in `gpu_test.ts`?

dneto0 commented on src/webgpu/gpu_test.ts:
```diff
@@ -245,6 +247,23 @@ export class GPUTestSubcaseBatchState extends SubcaseBatchState {
+  /** Skips this test case if the `langFeature` is *not* supported. */
+  requireLanguageFeatureOrSkipTestCase(langFeature: WGSLLanguageFeature) {
+    const lf = getGPU(this.recorder).wgslLanguageFeatures;
+    if (!(langFeature in lf)) {
```
Using      lf.has(langFeature) failed with an UNIMPLEMENTED exception.
This way of doing it properly skips or does not skip.

kainino0x replied:
wgslLanguageFeatures appears to be incorrectly implemented in dawn.node.
`langFeature in lf` should always return false because that's just checking for a key with a name equal to `langFeature` in the JS object `lf`. But that's just the interface to the set, so it only has keys like `has`, etc.
`lf.has(langFeature)` is the correct way.

Also we may want to skip if `wgslLanguageFeatures` doesn't exist (to make life easier for incomplete WebGPU implementations). In this case `lf` would be `undefined`.

dneto0 replied:
Thanks. fixed now. Tested with Chrome.
At least the "unsupported" test does trigger and detect the compilation failures.

kainino0x commented on src/webgpu/wgsl_language_feature.ts:
```diff
@@ -0,0 +1,12 @@
+/**
+ * @fileoverview Declares a type for each string tested against
+ * the WGSLLanguageFeatures property of the GPU object.
+ */
```
Unfortunately I don't think TSDoc supports file-level documentation. This is getting attached to `kKnownWGSLLanguageFeatures` (visible in vscode or `docs/tsdoc`).

Looking into this it seems like TSDoc is kind of dying and we should probably use JSDoc? I'll attempt this sometime.

kainino0x commented on src/webgpu/wgsl_language_feature.ts:
```diff
@@ -0,0 +1,12 @@
+/**
```
Usually we just add new things like this to `capability_info.ts`. This will stay small enough that I'd say just put it there.

kainino0x commented on src/resources/cache/hashes.json:
```diff
@@ -1,106 +1,106 @@
 {
```
It makes sense that these hash files change but the binary files should not have. See the bold thing here: https://github.com/gpuweb/cts/blob/main/docs/case_cache.md#file-cache

If this is all it took to trigger this - and I should have realized it would be - then I think we're going to need to fix that issue sooner rather than later. @ben-clayton

ben-clayton replied:
https://github.com/gpuweb/cts/pull/3187

kainino0x commented on src/webgpu/gpu_test.ts:
```diff
@@ -245,6 +247,22 @@ export class GPUTestSubcaseBatchState extends SubcaseBatchState {
   }
+
+  /** Skips this test case if the `langFeature` is *not* supported. */
+  requireLanguageFeatureOrSkipTestCase(langFeature: WGSLLanguageFeature) {
```
On second thought (re: my comment from October), this should probably be on GPUTest and not GPUTestSubcaseBatchState.

The reason for this is that, unlike selectDeviceOrSkipTestCase, this doesn't have to be done before getting a GPUDevice - wgslLanguageFeatures is static.

kainino0x replied:
GPUTestBase, actually, not GPUTest. It can go next to skipIfTextureFormatNotSupported.

dneto0 commented on src/webgpu/wgsl_language_feature.ts:
```diff
@@ -0,0 +1,12 @@
+  'readonly_and_readwrite_storage_textures',
+  'packed_4x8_integer_dot_product',
+  'unrestricted_pointer_parameters',
+];
```
committed

kainino0x reviewed (commented):
LGTM, but please undo all of the .bin changes - none of those files actual (decompressed) content should have changed due to this patch.

dneto0 commented:
> LGTM, but please undo all of the .bin changes - none of those files actual (decompressed) content should have changed due to this patch.

I've squashed the primary content, removed the .bin changes, and updated the hashes.
(The CI failed before I updated the caches)

dneto0 commented:
> > LGTM, but please undo all of the .bin changes - none of those files actual (decompressed) content should have changed due to this patch.
> 
> I've squashed the primary content, removed the .bin changes, and updated the hashes. (The CI failed before I updated the caches)

I'm installing node v20 to see if that makes a difference
